### ISSUE TITLE
Support adding more Java Flight Recorder data with `jfrsync`

### ISSUE THREAD

**Issue by charlesverdad** [async-profiler/async-profiler#1076]
### Describe the feature

Async Profiler with `jfrsync` flag supports syncing data from **Java Flight Recorder**. However, when opened in **JDK Mission Control**, some key tabs, namely Memory, Method Profiling, and Exceptions seem to be empty.

The images below compare a profile created by the JFR vs one created by Async-Profiler v3.0 with the `jfrsync=default` flag. Notice that the ones created by Async-Profiler is empty.

#### Java Flight Recorder
<img width="1015" alt="Screenshot 2024-12-09 at 2 07 29 PM" src="https://github.com/user-attachments/assets/6a2a8577-0ba0-49d9-a72a-6de586b3f9c4">

#### Async-Profiler 3.0

<img width="1015" alt="Screenshot 2024-12-09 at 2 20 59 PM" src="https://github.com/user-attachments/assets/212147f2-9da6-4f0c-8062-10e11984b87f">

I see in the v3.0 that we started supporting GC heap allocations. Is it possible to support the other ones as well?

### Use Case

I'm trying to get my team to use Async Profiler but they want to retain Java Flight Recorder because of the additional data.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

**Comment by apangin** [comment]
I don't see an issue you've described. Can you please clarify how exactly you start profiler? What options do you use? What is the JVM version and JMC version?

In `jfrsync` mode, async-profiler runs in parallel with JFR, it collects exactly the same evens as would JFR collect on its own, except that execution samples come from async-profiler.

**Comment by charlesverdad** [comment]
I'm using async-profiler as a java agent with this flag:

```
-agentpath:/path/to/libasyncProfiler.so=start,event=itimer,lock=10ms,alloc=1m,file=/tmp/%t-%{HOSTNAME}.jfr,loop=60s,jfrsync=default
```

OpenJDK version 17
JMC version: JDK Mission Control 8.3.1+b05


>  it collects exactly the same evens as would JFR collect on its own, except that execution samples come from async-profiler.

That's what I was expecting as well, but upgrading from 2.8 to 3.0 now gives me data in the "Garbage collections" tab. It was always empty before, and I attribute it to this [commit](https://github.com/async-profiler/async-profiler/commit/6c9b71594ca3bff3824aade538303570623a1011), so I thought we just need to support more types.

Perhaps it's because I'm using `loop=60s` and that isn't long enough? 🤔

**Comment by apangin** [comment]
I still see no issues. Do you have a problematic .jfr recording you can share?

**Comment by charlesverdad** [comment]
Thanks for looking into this @apangin .

I created a [test project](https://github.com/charlesverdad/jfr_vs_ap/tree/main) and I can confirm that everything works as expected in my local environment _and_ in a local docker environment, so this is not a problem with Async-Profiler.

Somehow in our ECS setup the memory data do not appear if I use `jfrsync` flag along with `event=itimer,lock,alloc`.

```
-agentpath:/path/to/libasyncProfiler.so=start,event=itimer,lock=10ms,alloc=1m,file=/tmp/%t-%{HOSTNAME}.jfr,loop=60s,jfrsync=default
```

But it *works* if I drop the `lock` and `alloc` event flags like so:

```
-agentpath:/path/to/libasyncProfiler.so=start,event=itimer,file=/tmp/%t-%{HOSTNAME}.jfr,loop=60s,jfrsync=default
```

My initial thoughts were it could be related to permissions or app armour config (i.e. ptrace is explicitly denied. But somehow it works too if I just don't use the jfrsync flag.

```
-agentpath:/path/to/libasyncProfiler.so=start,event=itimer,lock=10ms,alloc=1m,file=/tmp/%t-%{HOSTNAME}.jfr,loop=60s
```

**Comment by apangin** [comment]
Thank you for the update.
What does `jfr summary` print for the problematic profile?

**Comment by charlesverdad** [comment]
Here's the summary of the problematic JFR:

```

 Version: 2.1
 Chunks: 1
 Start: 2024-12-05 05:18:22 (UTC)
 Duration: 36 s

 Event Type                                                          Count  Size (bytes) 
=========================================================================================
 jdk.ModuleExport                                                     2012         21962
 jdk.ClassLoaderStatistics                                             614         15750
 jdk.BooleanFlag                                                       538         16446
 jdk.ActiveSetting                                                     366         11331
 jdk.ThreadSleep                                                       363          6174
 jdk.ThreadAllocationStatistics                                        339          4382
 jdk.ThreadCPULoad                                                     208          3625
 jdk.LongFlag                                                          198          6578
 jdk.ModuleRequire                                                     160          1600
 jdk.UnsignedLongFlag                                                  144          4934
 jdk.InitialEnvironmentVariable                                        101         15420
 jdk.JavaMonitorWait                                                   100          2948
 jdk.CheckPoint                                                         88        138769
 jdk.NativeLibrary                                                      58          4882
 io.lettuce.core.event.metrics.JfrCommandLatency                        56          2692
 jdk.InitialSecurityProperty                                            46          3674
 jdk.InitialSystemProperty                                              40          2458
 jdk.CPULoad                                                            36           720
 jdk.JavaThreadStatistics                                               36           576
 jdk.ClassLoadingStatistics                                             36           432
 jdk.CompilerStatistics                                                 36          1188
 jdk.ExceptionStatistics                                                36           540
 jdk.StringFlag                                                         31           971
 jdk.ThreadStart                                                        22           331
 jdk.ThreadEnd                                                          19           228
 jdk.NetworkUtilization                                                 12           177
 io.lettuce.core.event.connection.JfrConnectionActivatedEvent           12          1416
 io.lettuce.core.event.connection.JfrConnectedEvent                     12          1477
 io.lettuce.core.event.connection.JfrConnectionCreatedEvent             12           442
 io.lettuce.core.event.connection.JfrConnectEvent                       12           323
 io.lettuce.core.event.connection.JfrDisconnectedEvent                  12           599
 io.lettuce.core.event.connection.JfrConnectionDeactivatedEvent         12           599
 jdk.DoubleFlag                                                         11           443
 jdk.Deoptimization                                                      9           242
 jdk.SocketRead                                                          8           544
 jdk.DirectBufferStatistics                                              8           216
 jdk.UnsignedIntFlag                                                     7           240
 jdk.CodeCacheStatistics                                                 6           202
 jdk.IntFlag                                                             6           220
 io.lettuce.core.cluster.event.JfrTopologyRefreshEvent                   4          1188
 jdk.ThreadContextSwitchRate                                             4            48
 jdk.SymbolTableStatistics                                               4           156
 jdk.StringTableStatistics                                               4           156
 jdk.PlaceholderTableStatistics                                          4           136
 jdk.LoaderConstraintsTableStatistics                                    4           136
 jdk.ProtectionDomainCacheTableStatistics                                4           144
 io.lettuce.core.event.metrics.JfrCommandLatencyEvent                    3            42
 jdk.ContainerCPUUsage                                                   2            46
 jdk.ContainerCPUThrottling                                              2            30
 jdk.ContainerMemoryUsage                                                2            46
 jdk.ContainerIOUsage                                                    2            34
 jdk.PhysicalMemory                                                      2            36
 jdk.ThreadDump                                                          2        348470
 jdk.CodeSweeperStatistics                                               2            48
 jdk.GCConfiguration                                                     2            52
 jdk.Metadata                                                            1        101179
 jdk.ContainerConfiguration                                              1            57
 jdk.JVMInformation                                                      1          2369
 jdk.OSInformation                                                       1           233
 jdk.VirtualizationInformation                                           1            28
 jdk.CPUInformation                                                      1          1448
 jdk.CPUTimeStampCounter                                                 1            20
 jdk.CompilerConfiguration                                               1            10
 jdk.CodeCacheConfiguration                                              1            46
 jdk.CodeSweeperConfiguration                                            1            13
 jdk.GCSurvivorConfiguration                                             1            11
 jdk.GCTLABConfiguration                                                 1            13
 jdk.GCHeapConfiguration                                                 1            28
 jdk.YoungGenerationConfiguration                                        1            22
 jdk.ActiveRecording                                                     1           114
 jdk.ThreadPark                                                          0             0
 jdk.JavaMonitorEnter                                                    0             0
 jdk.JavaMonitorInflate                                                  0             0
 jdk.SyncOnValueBasedClass                                               0             0
 jdk.BiasedLockRevocation                                                0             0
 jdk.BiasedLockSelfRevocation                                            0             0
 jdk.BiasedLockClassRevocation                                           0             0
 jdk.ReservedStackActivation                                             0             0
 jdk.ClassLoad                                                           0             0
 jdk.ClassDefine                                                         0             0
 jdk.ClassRedefinition                                                   0             0
 jdk.RedefineClasses                                                     0             0
 jdk.RetransformClasses                                                  0             0
 jdk.ClassUnload                                                         0             0
 jdk.IntFlagChanged                                                      0             0
 jdk.UnsignedIntFlagChanged                                              0             0
 jdk.LongFlagChanged                                                     0             0
 jdk.UnsignedLongFlagChanged                                             0             0
 jdk.DoubleFlagChanged                                                   0             0
 jdk.BooleanFlagChanged                                                  0             0
 jdk.StringFlagChanged                                                   0             0
 jdk.GCHeapSummary                                                       0             0
 jdk.MetaspaceSummary                                                    0             0
 jdk.MetaspaceGCThreshold                                                0             0
 jdk.MetaspaceAllocationFailure                                          0             0
 jdk.MetaspaceOOM                                                        0             0
 jdk.MetaspaceChunkFreeListSummary                                       0             0
 jdk.PSHeapSummary                                                       0             0
 jdk.G1HeapSummary                                                       0             0
 jdk.GarbageCollection                                                   0             0
 jdk.SystemGC                                                            0             0
 jdk.ParallelOldGarbageCollection                                        0             0
 jdk.YoungGarbageCollection                                              0             0
 jdk.OldGarbageCollection                                                0             0
 jdk.G1GarbageCollection                                                 0             0
 jdk.G1MMU                                                               0             0
 jdk.EvacuationInformation                                               0             0
 jdk.GCReferenceStatistics                                               0             0
 jdk.ObjectCountAfterGC                                                  0             0
 jdk.G1EvacuationYoungStatistics                                         0             0
 jdk.G1EvacuationOldStatistics                                           0             0
 jdk.G1BasicIHOP                                                         0             0
 jdk.G1AdaptiveIHOP                                                      0             0
 jdk.PromoteObjectInNewPLAB                                              0             0
 jdk.PromoteObjectOutsidePLAB                                            0             0
 jdk.PromotionFailed                                                     0             0
 jdk.EvacuationFailed                                                    0             0
 jdk.ConcurrentModeFailure                                               0             0
 jdk.GCCPUTime                                                           0             0
 jdk.GCPhasePause                                                        0             0
 jdk.GCPhasePauseLevel1                                                  0             0
 jdk.GCPhasePauseLevel2                                                  0             0
 jdk.GCPhasePauseLevel3                                                  0             0
 jdk.GCPhasePauseLevel4                                                  0             0
 jdk.GCPhaseConcurrent                                                   0             0
 jdk.GCPhaseConcurrentLevel1                                             0             0
 jdk.GCPhaseParallel                                                     0             0
 jdk.AllocationRequiringGC                                               0             0
 jdk.TenuringDistribution                                                0             0
 jdk.G1HeapRegionTypeChange                                              0             0
 jdk.JITRestart                                                          0             0
 jdk.Compilation                                                         0             0
 jdk.CompilerPhase                                                       0             0
 jdk.CompilationFailure                                                  0             0
 jdk.CompilerInlining                                                    0             0
 jdk.SweepCodeCache                                                      0             0
 jdk.CodeCacheFull                                                       0             0
 jdk.SafepointBegin                                                      0             0
 jdk.SafepointStateSynchronization                                       0             0
 jdk.SafepointCleanup                                                    0             0
 jdk.SafepointCleanupTask                                                0             0
 jdk.SafepointEnd                                                        0             0
 jdk.ExecuteVMOperation                                                  0             0
 jdk.Shutdown                                                            0             0
 jdk.ObjectAllocationInNewTLAB                                           0             0
 jdk.ObjectAllocationOutsideTLAB                                         0             0
 jdk.ObjectAllocationSample                                              0             0
 jdk.OldObjectSample                                                     0             0
 jdk.DumpReason                                                          0             0
 jdk.DataLoss                                                            0             0
 jdk.SystemProcess                                                       0             0
 org.springframework.core.metrics.jfr.FlightRecorderStartupEvent         0             0
 jdk.ExecutionSample                                                     0             0
 jdk.NativeMethodSample                                                  0             0
 jdk.ObjectCount                                                         0             0
 jdk.G1HeapRegionInformation                                             0             0
 jdk.ZAllocationStall                                                    0             0
 jdk.ZPageAllocation                                                     0             0
 jdk.ZRelocationSet                                                      0             0
 jdk.ZRelocationSetGroup                                                 0             0
 jdk.ZStatisticsCounter                                                  0             0
 jdk.ZStatisticsSampler                                                  0             0
 jdk.ZThreadPhase                                                        0             0
 jdk.ZUncommit                                                           0             0
 jdk.ZUnmap                                                              0             0
 jdk.ShenandoahHeapRegionStateChange                                     0             0
 jdk.ShenandoahHeapRegionInformation                                     0             0
 jdk.Flush                                                               0             0
 jdk.HeapDump                                                            0             0
 jdk.GCLocker                                                            0             0
 jdk.FileForce                                                           0             0
 jdk.FileRead                                                            0             0
 jdk.FileWrite                                                           0             0
 jdk.SocketWrite                                                         0             0
 jdk.JavaExceptionThrow                                                  0             0
 jdk.JavaErrorThrow                                                      0             0
 jdk.Deserialization                                                     0             0
 jdk.ProcessStart                                                        0             0
 jdk.SecurityPropertyModification                                        0             0
 jdk.SecurityProviderService                                             0             0
 jdk.TLSHandshake                                                        0             0
 jdk.X509Certificate                                                     0             0
 jdk.X509Validation                                                      0             0
```


When compared to the non-problematic one (using JFR), I see that the following events are missing:
```
jdk.OldObjectSample
jdk.SystemProcess
jdk.ExecuteVMOperation
jdk.MetaspaceSummary
jdk.GCPhasePauseLevel1
jdk.ThreadPark
jdk.GCPhaseConcurrent
org.springframework.core.metrics.jfr.FlightRecorderStartupEvent
jdk.JavaErrorThrow
jdk.MetaspaceAllocationFailure
jdk.NativeMethodSample
jdk.GarbageCollection
jdk.GCHeapSummary
jdk.ObjectAllocationSample
jdk.MetaspaceGCThreshold
jdk.MetaspaceChunkFreeListSummary
jdk.GCReferenceStatistics
jdk.JavaMonitorEnter
jdk.GCPhasePause
jdk.GCPhaseParallel
jdk.ExecutionSample
jdk.GCPhaseConcurrentLevel1
```

**Comment by apangin** [comment]
OK, thanks. It seems that you've got only JFR part of the recording with `jfrsync` option, there is no async-profiler chunk - apparently because copying has failed. Async-profiler uses `sendfile` syscall to merge files, this must have failed for some reason.

1. What is Linux kernel version?
2. What is the filesystem that .jfr recordings are written to?
3. Will you be able to test a custom version from a branch to verify my hypothesis?

**Comment by charlesverdad** [comment]
1. Kernel version: `5.15.0-1072-aws`
2. I'm using a local docker volume, which should be mapped to the host's /var/lib/docker/volumes , and is in turn an xfs filesystem
```
findmnt /var/lib/docker
TARGET          SOURCE       FSTYPE OPTIONS
/var/lib/docker /dev/nvme1n1 xfs    rw,relatime,attr2,inode64,logbufs=8,logbsize=32k,usrquota,prjquota
```
3. > Will you be able to test a custom version from a branch to verify my hypothesis?
Yes, I should be able to use a custom version. I can easily plug in and replace the `*.so` file I'm using.

**Comment by charlesverdad** [comment]
Hey @apangin , following your hypothesis I went to my instance and checked the `sendfile` syscall. It seems to be succeeding properly 🤔 Despite that, I see no data in the "Method Profiling" tab.

I ran the command below inside my host machine, where 5125 is the PID of the container running the java app.
```
~$ sudo nsenter --target 5125 --pid --mount -- ps -T -p 138 | grep Async-profiler
    138    2659 ?        00:00:00 Async-profiler
~$ sudo nsenter --target 5125 --pid -- strace -p 2659 -e trace=sendfile
strace: Process 2659 attached
sendfile(10, 1436, [0] => [164165], 164165) = 164165
+++ exited with 0 +++
```
I can provide more of the `strace` data if needed.

Looking at the specific events I'm missing, `jdk.NativeMethodSample` and `jdk.ExecutionSample`, I suspect it is going through [this block](https://github.com/async-profiler/async-profiler/blob/d8f1b9c15a224bfbd880fe7cc895023934fc286a/src/helper/one/profiler/JfrSync.java#L86-L89) which disables recording them. I'm not sure what the condition `(eventMask & 1) != 0` means, but the flags I'm using `event=itimer,jfrsync=default` I think it shouldn't call those disable methods 🤔 

Does this disable call mean to disable JFR from listening to these events because async-profiler is going to measure them directly (but somehow fails in my setup)?

**Comment by apangin** [comment]
> Does this disable call mean to disable JFR from listening to these events because async-profiler is going to measure them directly

Yes: in `jfrsync mode`, async-profiler disables collection of certain events by the JDK Flight Recorder that async-profiler collects on its own. The main problem here is that `jfr summary` prints `Chunks: 1`, while there should be 2 or more chunks.

I've changed the way how recordings are merged - try #1083. Can you try this code?
Binaries for this PR are available at https://github.com/async-profiler/async-profiler/actions/runs/12374480872

**Comment by charlesverdad** [comment]
Thanks for the explanation, and for that fix! I tried the custom build and it didn't _seem_ to work.

However, I think I found the root cause of my problem.

I noticed that with `jfrsync` flag, a temporary file is created and is written to again. Presumably this temporary one contains the async profiler dump and is later appended to with JFR data.  This is problematic because in my ECS setup I have a process (sidecar container) that listens to the created files so that it can upload them to S3. Since it also deletes the file after uploading, I am then left with only half the data.

When I kill my sidecar container I can see that `jfr summary` on the file in disk now returns `Chunks: 2`, but the ones uploaded to S3 all have `Chunks:1`. 

I think I can easily solve this by adding a delay, or removing the delete logic.

(I tried this both with the official released 3.0 and the one in https://github.com/async-profiler/async-profiler/pull/1083 and they both worked fine.)
